### PR TITLE
Support shared cluster billing in the deployer

### PR DESCRIPTION
**Pull request by pnasrat**
Extract billing module and refactor into a class
Use datetime parameter typer argument to validate
Move to use pandas for bigquery data
Adds cluster.yaml schema field for tenancy type (dedicated, shared) Create unit tests for billing
Wiring and initial code for shared cluster costs using prometheus importer

### CODE REVIEW

**Comment by pnasrat** [comment]
This is a draft/WIP pushed for structural feedback only. Validate fails as (suprising to me)  adding a default value to the schema doesn't actually make it default set on parsing cluster.yaml 

It also looks like I need to add more mocking so that we don't try to use actual credentials from the CI test failure

**Comment by pnasrat** [comment]
Local testing. 

```
 pytest tests/test_billing.py 
============================================================================================= test session starts =============================================================================================
platform linux -- Python 3.9.16, pytest-7.3.1, pluggy-1.0.0
rootdir: /home/pnasrat/workspace/src/github.com/2i2c-org/wt/billing-shared-cluster
configfile: pyproject.toml
plugins: typeguard-4.0.0, asyncio-0.21.0
asyncio: mode=auto
collected 2 items                                                                                                                                                                                             

tests/test_billing.py ..                                                                                                                                                                                [100%]

============================================================================================== 2 passed in 1.24s ==============================================================================================
```

I'm also wondering if there is a reason we're not using junitxml output for tests in CI along with something like test reporter action https://github.com/marketplace/actions/test-reporter

File: deployer/billing/cost_importer.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    return result
+
+
+# TODO(pnasrat): wire in prometheus values
+def get_shared_hub_costs(cluster, start_month, end_month):
+    costs = get_dedicated_cluster_costs(cluster, start_month, end_month)
+    prom_importer = PrometheusUtilizationImporter(cluster)
+    utilization = prom_importer.get_costs(start_month, end_month)
+    totals = costs.merge(utilization, how="left", on="month")
+    totals.iloc[:, 2:].multiply(totals.iloc[:, 1].astype(float), axis=0)
```
Comment (yuvipanda) [inline deployer/billing/cost_importer.py:156]: If I understand this correctly, this is the core of computing shared costs - we get these to be two pd dataframes and use pandas to do appropriate “math”. We can extend this in complexity in the future by adding appropriate extra columns, etc. I don’t know a lot of pandas, but the community is large enough that i believe we can make it work.


If my assumption of what this approach is is correct, then i am a big fan! I vastly prefer this to the lists & dict stuff i had written for dedicated clusters earlier!

i will provide a more detailed review monday, but at a general level, +1!

Comment (pnasrat) [reply deployer/billing/cost_importer.py:156]: Yes at a high level I'm trying to make it more portable going forward using pandas. It allows us to replicate the approach from the spreadsheets programatically and is extendible as  you say. Some of the slowgoing has been me trying to figure out good practices in libraries I'm less familiar with.

Comment (yuvipanda) [reply deployer/billing/cost_importer.py:156]: @pnasrat awesome, i love the approach!!!!

**Comment by pnasrat** [comment]
Sample output

https://docs.google.com/spreadsheets/d/1T6BJkAhlLtK4U6YB5pVPQKpVej_La0pvG3AlzDOAJ60/edit#gid=0

**Comment by pnasrat** [comment]
@yuvipanda would you be able to review this some time over the next two weeks while I'm on vacation then I will address feedback and merge prior to Q2 billing

**Comment by yuvipanda** [comment]
Yes I'll provide a response to this coming week.

**Review by yuvipanda — APPROVED** [review]
Generally lgtm, with some minor style nits.

File: deployer/billing/cost_importer.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        # TODO(pnasrat): Pass client in or encapsulate in an class?
+        self.client = bigquery.Client()
+
+        # TODO(pnasrat): we are using GCP project name not cluster name.
+        self.cluster_project_name = cluster["gcp"]["project"]
+        self.bq = cluster["gcp"]["billing"]["bigquery"]
+
+        self.table_name = f'{self.bq["project"]}.{self.bq["dataset"]}.gcp_billing_export_resource_v1_{self.bq["billing_id"].replace("-", "_")}'
+        # Make sure the table name only has alphanumeric characters, _ and -
+        assert re.match(r"^[a-zA-Z0-9._-]+$", self.table_name)
```
[the remark this answers is not in the document]
Comment (pnasrat) [reply deployer/billing/cost_importer.py]: What is the actual threat model here?

Can we even write to these tables - I don't imagine customers can write to GCP billing? 

Even if so I think this is the case where if we ensure the credentials only has BigQuery reader.

File: deployer/billing/cost_importer.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    """Bigquery Cost Importer for GCP clusters
+
+    Args:
+        cluster (dict): parsed cluster.yaml
+    """
+
+    def __init__(self, cluster: dict, query=None):
+        self.cluster = cluster
+
+        # TODO(pnasrat): Pass client in or encapsulate in an class?
```
[the remark this answers is not in the document]
Comment (pnasrat) [reply deployer/billing/cost_importer.py]: This is an infrequent task so probalby no need to optimize at this moment. I'll remove the comment

File: deployer/billing/cost_importer.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+        if "support" in df:
+            df["support_combined"] = df["support_combined"] + df["support"]
+
+        if "kube-system" in df:
+            df["support_combined"] = df["support_combined"] + df["kube-system"]
+        return df.drop(columns=["support", "kube-system"], errors="ignore")
+
+    def combine_internal_costs(self, df):
+        internal = [
+            "utexas-demo",
```
[the remark this answers is not in the document]
Comment (pnasrat) [reply deployer/billing/cost_importer.py:168]: While I agree we should have an external source than cluster.yaml for billing that correctly models contracts, costs we take on, etc that's very much beyond the scope of this automate what we did in Q1 for the shared cluster. I'll add a comment.

File: deployer/billing/cost_importer.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
+import re
```
[the remark this answers is not in the document]
Comment (pnasrat) [reply deployer/billing/cost_importer.py:1]: @yuvipanda I'm open to other naming,  we aren't really using standard forms to work with data (eg an ORM with an idea of repository). I guess I was thinking of importing from external datasource here in the workflow import, clean, transform (cf data science practices).

**Comment by pnasrat** [comment]
@yuvipanda please take another look

File: deployer/billing/cost_importer.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+        df["support_combined"] = 0.0
+        if "support" in df:
+            df["support_combined"] = df["support_combined"] + df["support"]
+
+        if "kube-system" in df:
+            df["support_combined"] = df["support_combined"] + df["kube-system"]
+        return df.drop(columns=["support", "kube-system"], errors="ignore")
+
+    def combine_internal_costs(self, df):
+        # This list was manually created during the Q1 2023 billing consolidation work.
```
Comment (yuvipanda) [inline deployer/billing/cost_importer.py:175]:
```suggestion
        # FIXME: This list was manually created during the Q1 2023 billing consolidation work.
```

File: deployer/billing/cost_importer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yuvipanda) [reply deployer/billing/cost_importer.py:178]: Makes sense, I just added a suggestion for `FIXME`.

File: deployer/billing/cost_importer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (yuvipanda) [reply deployer/billing/cost_importer.py:1]: yeah that's fair enough. Given there is both a cost and utilization importer here, can we rename the file to `importer.py`?

**Review by yuvipanda — APPROVED** [review]
Two extremely minor style suggestions but good to merge after that. Thank you for working on this, @pnasrat!

File: deployer/billing/cost_importer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yuvipanda) [reply deployer/billing/cost_importer.py]: The threat model for me is mostly about how this code *evolves* over time, and unanticipated additions in the future. String concatenation for making queries feels like something we should avoid at all costs, and though there is nothing specific here right now I'm worried that if we don't signpost this hard enough there's a possibility that this code will expand in the future in some way that can cause information leakage. It does feel *slightly* overkill for this situation (deployer is only run by us), but there's enough SQL Injection battle scars around that I feel better keeping this warning here.

**Comment by pnasrat** [comment]
Cancelled deploy as schemq change triggered unnecessary full deploy

**Pull request merged**